File: Client/models/TemplateDTO.ts
import { DataType } from '~/models/Enums/DataType';
import { PatchOption } from './Enums/PatchOption';

export interface TemplateGetDTO {
  name: string;
  attributes: TemplateAttributeGetDTO[];
  id: number;
}

export interface TemplateAttributeGetDTO {
  attributeId: number;
  typeId: number;
  name: string;
  type: string;
  dataType: DataType;
  featured: boolean;
  id: number;
  usesDefinedValues: boolean,
  usesDefinedUnits: boolean,
}

export interface TemplatePostDTO {
  name: string;
  attributes: TemplateAttributePostDTO[];
}

export interface TemplateAttributePostDTO {
  attributeId: number;
  featured: boolean;
}

export interface TemplatePatchDTO {
  id: number;
  name: string;
  attributes: TemplateAttributePatchDTO[];
}

export interface TemplateAttributePatchDTO {
  id: number | null;
  featured: boolean;
  attributeId: number;
  patchOption: PatchOption;
}

File: Server/AppAPI/1.0/TemplateDTO.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using AppAPI._1._0.Enums;
using DAL.Base.Entities;

namespace AppAPI._1._0
{
    #region GET

    public class TemplateGetDTO: DomainEntityId
    {
        public string Name { get; set; } = default!;

        public ICollection<TemplateAttributeGetDTO>? Attributes { get; set; }
    }

    public class TemplateAttributeGetDTO: DomainEntityId
    {
        public long AttributeId { get; set; }
        public long TypeId { get; set; }
        
        public string Name { get; set; } = default!;
        public string Type { get; set; } = default!;
        
        public AttributeDataType DataType { get; set; }
        
        public bool Featured { get; set; }
        
        public bool UsesDefinedValues { get; set; }
        public bool UsesDefinedUnits { get; set; }
    }

    #endregion

    #region POST

    public class TemplatePostDTO
    {
        [Required][MaxLength(100)]  public string Name { get; set; } = default!;

        [MaxLength(30)] public ICollection<TemplateAttributePostDTO> Attributes { get; set; } = default!;
    }

    public class TemplateAttributePostDTO
    {
        public bool Featured { get; set; }
        [Required] public long AttributeId { get; set; } = default!;
    }

    #endregion
    
    #region PATCH

    public class TemplatePatchDTO: DomainEntityId
    {
        [Required][MaxLength(100)]  public string Name { get; set; } = default!;
        
        [MaxLength(30)] public ICollection<TemplateAttributePatchDTO> Attributes { get; set; } = default!;
    }

    public class TemplateAttributePatchDTO
    {
        public long? Id { get; set; }
        
        public PatchOption PatchOption { get; set; }
        
        public bool Featured { get; set; }
        
        [Required] public long AttributeId { get; set; } = default!;
    }

    #endregion
}